apis/kubedb/v1/mongodb_helpers.go

```diff
@@ -0,0 +1,1024 @@
+	return m.ServiceLabels(StatsServiceAlias, map[string]string{kubedb.LabelRole: kubedb.RoleStats})
+}
+
+// StorageType = Durable
+// storageEngine = wiredTiger
+// TerminationPolicy = Delete
+// SSLMode = disabled
+// clusterAuthMode = keyFile if sslMode is disabled or allowSSL.  & x509 otherwise
+//
+// podTemplate.Spec.ServiceAccountName = DB_NAME
+// set mongos lifecycle command, to shut down the db before stopping
+// it sets default ReadinessProbe, livelinessProbe, affinity, ResourceLimits & securityContext
+// then set TLSDefaults & monitor Defaults
+
+func (m *MongoDB) SetDefaults(mgVersion *v1alpha1.MongoDBVersion, topology *core_util.Topology) {
+	if m == nil {
+		return
+	}
+
+	if m.Spec.StorageType == "" {
+		m.Spec.StorageType = StorageTypeDurable
+	}
+	if m.Spec.StorageEngine == "" {
+		m.Spec.StorageEngine = StorageEngineWiredTiger
+	}
+	if m.Spec.TerminationPolicy == "" {
+		m.Spec.TerminationPolicy = TerminationPolicyDelete
+	}
+
+	if m.Spec.SSLMode == "" {
+		m.Spec.SSLMode = SSLModeDisabled
+	}
+
+	if (m.Spec.ReplicaSet != nil || m.Spec.ShardTopology != nil) && m.Spec.ClusterAuthMode == "" {
+		if m.Spec.SSLMode == SSLModeDisabled || m.Spec.SSLMode == SSLModeAllowSSL {
+			m.Spec.ClusterAuthMode = ClusterAuthModeKeyFile
+		} else {
+			m.Spec.ClusterAuthMode = ClusterAuthModeX509
+		}
+	}
+
+	if m.Spec.ShardTopology != nil {
+		m.setPodTemplateDefaultValues(&m.Spec.ShardTopology.Mongos.PodTemplate, mgVersion)
+		m.setPodTemplateDefaultValues(&m.Spec.ShardTopology.Shard.PodTemplate, mgVersion)
+		m.setPodTemplateDefaultValues(&m.Spec.ShardTopology.ConfigServer.PodTemplate, mgVersion)
+	} else {
+		if m.Spec.Replicas == nil {
+			m.Spec.Replicas = pointer.Int32P(1)
+		}
+
+		m.setPodTemplateDefaultValues(m.Spec.PodTemplate, mgVersion)
+
+		if m.Spec.ReplicaSet != nil {
+			if m.Spec.Coordinator.SecurityContext == nil {
+				m.Spec.Coordinator.SecurityContext = &core.SecurityContext{}
+			}
+			m.assignDefaultContainerSecurityContext(mgVersion, m.Spec.Coordinator.SecurityContext) // modeDetector container
+		}
+	}
+
+	if m.Spec.Arbiter != nil {
+		m.setPodTemplateDefaultValues(&m.Spec.Arbiter.PodTemplate, mgVersion, true)
+	}
+	if m.Spec.Hidden != nil {
+		m.setPodTemplateDefaultValues(&m.Spec.Hidden.PodTemplate, mgVersion)
+	}
+
+	m.SetTLSDefaults()
+	m.SetHealthCheckerDefaults()
+	m.Spec.Monitor.SetDefaults()
+	if m.Spec.Monitor != nil && m.Spec.Monitor.Prometheus != nil {
+		if m.Spec.Monitor.Prometheus.Exporter.SecurityContext.RunAsUser == nil {
+			m.Spec.Monitor.Prometheus.Exporter.SecurityContext.RunAsUser = mgVersion.Spec.SecurityContext.RunAsUser
+		}
+		if m.Spec.Monitor.Prometheus.Exporter.SecurityContext.RunAsGroup == nil {
+			m.Spec.Monitor.Prometheus.Exporter.SecurityContext.RunAsGroup = mgVersion.Spec.SecurityContext.RunAsGroup
+		}
+	}
+}
+
+func (m *MongoDB) setPodTemplateDefaultValues(podTemplate *ofstv2.PodTemplateSpec, mgVersion *v1alpha1.MongoDBVersion, isArbiter ...bool) {
```

Comment:
This Function panics for some reason. Which is failing unit tests.

Reply:
checking

Reply:
done